lib/rand/securesource.go

```diff
@@ -47,18 +46,18 @@ func (s *secureSource) Read(p []byte) (int, error) {
 }
 
 func (s *secureSource) Uint64() uint64 {
-	var buf [8]byte
-
 	// Read eight bytes of entropy from the buffered, secure random number
 	// generator. The buffered reader isn't concurrency safe, so we lock
 	// around that.
 	s.mut.Lock()
-	_, err := io.ReadFull(s.rd, buf[:])
-	s.mut.Unlock()
+	defer s.mut.Unlock()
+
+	buf, err := s.rd.Peek(8) // Peek fills the buffer if needed.
```

Comment:
While this is clever, could we not accomplish approximately the same thing by moving the `[8]byte` into the `secureSource` and the not have to depend on the reader being precisely a bufio.Reader and precisely how Peek/Discard word?

Reply:
Yes, at a minor cost in throughput:

```
name            old time/op    new time/op    delta
SecureSource-8    47.8ns ± 1%    51.7ns ± 3%  +8.21%  (p=0.000 n=18+10)
String-8          1.83µs ± 1%    1.95µs ± 1%  +6.62%  (p=0.000 n=10+10)
```

It's your call. To be sure, those comments are there to make some promises in the Peek/Discard descriptions explicit, they're not implementation details.

Reply:
Never mind. I just realized that Peek only ever does one Read, so this panics on intermittent failures. Pushed a version with a buffer in the secureSource.